In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

In [2]:
df=pd.read_csv("train.csv")

In [3]:
df.head(10)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
0,1,CA-2017-152156,08/11/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600
1,2,CA-2017-152156,08/11/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400
2,3,CA-2017-138688,12/06/2017,16/06/2017,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200
3,4,US-2016-108966,11/10/2016,18/10/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775
4,5,US-2016-108966,11/10/2016,18/10/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680
5,6,CA-2015-115812,09/06/2015,14/06/2015,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600
6,7,CA-2015-115812,09/06/2015,14/06/2015,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800
7,8,CA-2015-115812,09/06/2015,14/06/2015,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520
8,9,CA-2015-115812,09/06/2015,14/06/2015,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040
9,10,CA-2015-115812,09/06/2015,14/06/2015,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032.0,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000


In [4]:
df["Order Date"] = pd.to_datetime(df["Order Date"], format="%d/%m/%Y")
df["Ship Date"] = pd.to_datetime(df["Ship Date"], format="%d/%m/%Y")

In [5]:
df["Order Year"] = df["Order Date"].dt.year
df["Order Month"] = df["Order Date"].dt.month
df["Order Week"] = df["Order Date"].dt.isocalendar().week.astype(int)
df["Order Day of Week"] = df["Order Date"].dt.day_name()
df["Order Quarter"] = df["Order Date"].dt.quarter

# Function to determine season
def get_season(month):
    if month in [12, 1, 2]:
        return "Winter"
    elif month in [3, 4, 5]:
        return "Spring"
    elif month in [6, 7, 8]:
        return "Summer"
    else:
        return "Autumn"

df["Order Season"] = df["Order Month"].apply(get_season)

In [6]:
df["Ship Year"] = df["Ship Date"].dt.year
df["Ship Month"] = df["Ship Date"].dt.month
df["Ship Week"] = df["Ship Date"].dt.isocalendar().week.astype(int)
df["Ship Day of Week"] = df["Ship Date"].dt.day_name()
df["Ship Quarter"] = df["Ship Date"].dt.quarter
df["Ship Season"] = df["Ship Month"].apply(get_season)

# Display sample
print(df.head())

   Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2017-152156 2017-11-08 2017-11-11    Second Class    CG-12520   
1       2  CA-2017-152156 2017-11-08 2017-11-11    Second Class    CG-12520   
2       3  CA-2017-138688 2017-06-12 2017-06-16    Second Class    DV-13045   
3       4  US-2016-108966 2016-10-11 2016-10-18  Standard Class    SO-20335   
4       5  US-2016-108966 2016-10-11 2016-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ... Order Week  \
0      Claire Gute   Consumer  United States        Henderson  ...         45   
1      Claire Gute   Consumer  United States        Henderson  ...         45   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...         24   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...         41   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...         41   

   Order Day of Week Order Quarter Ord

In [7]:
df.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Country               0
City                  0
State                 0
Postal Code          11
Region                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Order Year            0
Order Month           0
Order Week            0
Order Day of Week     0
Order Quarter         0
Order Season          0
Ship Year             0
Ship Month            0
Ship Week             0
Ship Day of Week      0
Ship Quarter          0
Ship Season           0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.dtypes

Row ID                        int64
Order ID                     object
Order Date           datetime64[ns]
Ship Date            datetime64[ns]
Ship Mode                    object
Customer ID                  object
Customer Name                object
Segment                      object
Country                      object
City                         object
State                        object
Postal Code                 float64
Region                       object
Product ID                   object
Category                     object
Sub-Category                 object
Product Name                 object
Sales                       float64
Order Year                    int32
Order Month                   int32
Order Week                    int64
Order Day of Week            object
Order Quarter                 int32
Order Season                 object
Ship Year                     int32
Ship Month                    int32
Ship Week                     int64
Ship Day of Week            

In [ ]:
daily_sales = (df.groupby("Order Date")["Sales"]
                .sum()
                .reset_index()
                .sort_values("Order Date"))

print("Daily Sales")
print(daily_sales.head())

Daily Sales
  Order Date     Sales
0 2015-01-03    16.448
1 2015-01-04   288.060
2 2015-01-05    19.536
3 2015-01-06  4407.100
4 2015-01-07    87.158


In [12]:
weekly_sales = (daily_sales
    .set_index("Order Date")
    .resample("W")["Sales"]
    .sum()
    .reset_index())

print("\nWeekly Sales")
print(weekly_sales.head())


Weekly Sales
  Order Date     Sales
0 2015-01-04   304.508
1 2015-01-11  4619.108
2 2015-01-18  4130.533
3 2015-01-25  3092.544
4 2015-02-01  2527.914


In [13]:
monthly_sales = (daily_sales
    .set_index("Order Date")
    .resample("M")["Sales"]
    .sum()
    .reset_index())

print("\nMonthly Sales")
print(monthly_sales.head())


Monthly Sales
  Order Date      Sales
0 2015-01-31  14205.707
1 2015-02-28   4519.892
2 2015-03-31  55205.797
3 2015-04-30  27906.855
4 2015-05-31  23644.303


C:\Users\vasu mittal\AppData\Local\Temp\ipykernel_6772\2788439826.py:3: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  .resample("M")["Sales"]


In [18]:
# Total revenue by category
category_sales = (
    df.groupby("Category")["Sales"]
      .sum()
      .sort_values(ascending=False)
)

print(category_sales)

highest_category = category_sales.idxmax()
highest_revenue = category_sales.max()

print(f"\nHighest Revenue Category: {highest_category}")
print(f"Total Revenue: ${highest_revenue:,.2f}")



Category
Technology         827455.8730
Furniture          728658.5757
Office Supplies    705422.3340
Name: Sales, dtype: float64

Highest Revenue Category: Technology
Total Revenue: $827,455.87


In [19]:
# Extract year
df["Year"] = df["Order Date"].dt.year

# Annual sales by region
region_year_sales = (
    df.groupby(["Region", "Year"])["Sales"]
      .sum()
      .reset_index()
)

print(region_year_sales)

# Pivot for easier comparison
pivot = region_year_sales.pivot(
    index="Year",
    columns="Region",
    values="Sales"
)

print("\nAnnual Sales by Region")
print(pivot)

# Growth rate
growth = pivot.pct_change() * 100

print("\nYear-over-Year Growth (%)")
print(growth)

# Standard deviation of growth
consistency = growth.std().sort_values()

print("\nConsistency (Lower Std = More Consistent)")
print(consistency)


# Shipping duration
df["Shipping Days"] = (
    df["Ship Date"] - df["Order Date"]
).dt.days

# Overall average
overall_avg = df["Shipping Days"].mean()

print(f"Overall Average Shipping Time: {overall_avg:.2f} days")

# Regional averages
region_shipping = (
    df.groupby("Region")["Shipping Days"]
      .agg(["mean", "min", "max", "std"])
      .round(2)
)

print(region_shipping)


     Region  Year        Sales
0   Central  2015  102920.5206
1   Central  2016  102425.1724
2   Central  2017  145673.8800
3   Central  2018  141627.3402
4      East  2015  127652.8190
5      East  2016  153225.1830
6      East  2017  178511.5380
7      East  2018  210129.1860
8     South  2015  103374.9055
9     South  2016   70076.0825
10    South  2017   93535.9035
11    South  2018  122164.5675
12     West  2015  145907.9630
13     West  2016  133709.5675
14     West  2017  182471.2285
15     West  2018  248130.9255

Annual Sales by Region
Region      Central        East        South         West
Year                                                     
2015    102920.5206  127652.819  103374.9055  145907.9630
2016    102425.1724  153225.183   70076.0825  133709.5675
2017    145673.8800  178511.538   93535.9035  182471.2285
2018    141627.3402  210129.186  122164.5675  248130.9255

Year-over-Year Growth (%)
Region    Central       East      South       West
Year                   

In [21]:
# Shipping duration
df["Shipping Days"] = (
    df["Ship Date"] - df["Order Date"]
).dt.days

# Overall average
overall_avg = df["Shipping Days"].mean()

print(f"Overall Average Shipping Time: {overall_avg:.2f} days")

# Regional averages
region_shipping = (
    df.groupby("Region")["Shipping Days"]
      .agg(["mean", "min", "max", "std"])
      .round(2)
)

print(region_shipping)

Overall Average Shipping Time: 3.96 days
         mean  min  max   std
Region                       
Central  4.07    0    7  1.69
East     3.91    0    7  1.73
South    3.96    0    7  1.74
West     3.93    0    7  1.81


In [23]:
# Month and year
df["Month"] = df["Order Date"].dt.month
df["Month Name"] = df["Order Date"].dt.month_name()

# Monthly sales
monthly_sales = (
    df.groupby(["Year", "Month", "Month Name"])["Sales"]
      .sum()
      .reset_index()
)

# Pivot table
seasonality = monthly_sales.pivot(
    index="Month Name",
    columns="Year",
    values="Sales"
)

# Order months correctly
month_order = [
    "January","February","March","April","May","June",
    "July","August","September","October","November","December"
]

seasonality = seasonality.reindex(month_order)

print(seasonality)

# Average sales by month
avg_month = (
    monthly_sales.groupby("Month Name")["Sales"]
    .mean()
    .reindex(month_order)
    .sort_values(ascending=False)
)

avg_month.sort_values(ascending=False)

Year              2015        2016        2017         2018
Month Name                                                 
January     14205.7070  18066.9576  18542.4910   43476.4740
February     4519.8920  11951.4110  22978.8150   19920.9974
March       55205.7970  32339.3184  51165.0590   58863.4128
April       27906.8550  34154.4685  38679.7670   35541.9101
May         23644.3030  29959.5305  56656.9080   43825.9822
June        34322.9356  23599.3740  39724.4860   48190.7277
July        33781.5430  28608.2590  38320.7830   44825.1040
August      27117.5365  36818.3422  30542.2003   62837.8480
September   81623.5268  63133.6060  69193.3909   86152.8880
October     31453.3930  31011.7375  59583.0330   77448.1312
November    77907.6607  75249.3995  79066.4958  117938.1550
December    68167.0585  74543.6012  95739.1210   83030.3888


Month Name
November     87540.427750
December     80370.042375
September    75025.852925
October      49874.073675
March        49393.396800
August       39328.981750
May          38521.680925
June         36459.380825
July         36383.922250
April        34070.750150
January      23572.907400
February     14842.778850
Name: Sales, dtype: float64

Technology contributes the largest share of sales with $827,455.87.

The East region demonstrates the most stable and continuous year-over-year growth, making it the most consistent performing region.

The average shipping time across all orders is 3.96 days.
Regional observations:
Central has the longest average shipping time (4.07 days).
East has the shortest (3.91 days).
The difference between regions is only about 0.16 days, indicating shipping performance is fairly consistent nationwide.

Yes, the data exhibits clear seasonality. November and December consistently record the highest average sales across all years, indicating a strong increase in demand during the year-end holiday shopping season.On the other hand, January and February have the lowest average sales, indicating a post-holiday decline in customer purchases.